crates/shadowsocks/src/relay/udprelay/compat.rs

```diff
@@ -0,0 +1,88 @@
+use async_trait::async_trait;
+use std::{
+    io::Result,
+    net::SocketAddr,
+    ops::Deref,
+    task::{Context, Poll},
+};
+use tokio::io::ReadBuf;
+
+use crate::net::UdpSocket;
+
+/// a trait for datagram transport that wraps around a tokio `UdpSocket`
+#[async_trait]
+pub trait DatagramTransport: Send + Sync + std::fmt::Debug {
+    async fn recv(&self, buf: &mut [u8]) -> Result<usize>;
+    async fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, SocketAddr)>;
+
+    async fn send(&self, buf: &[u8]) -> Result<usize>;
+    async fn send_to(&self, buf: &[u8], target: SocketAddr) -> Result<usize>;
```

Comment:
Since we already have `poll_*` functions, these `async` wrappers could be added in an `DatagramTransportExt` trait.

Reply:
yes true, but thing is it won't work with ProxySocket without implementing all of these methods.

so i doubt spliting the Ext trait